lib/queue.js

```diff
@@ -1,256 +1,497 @@
-  } else {
-    Job.fromId(this, jobId, function (err, job) {
-      /* istanbul ignore if */
-      if (err) return cb(err);
-      self.jobs[jobId] = job;
-      return cb(err, job);
-    });
-  }
+  const promise = this._commandable().then(() => {
+    if (this.jobs.has(jobId)) {
+      return this.jobs.get(jobId);
+    }
+    return Job.fromId(this, jobId);
+  }).then((job) => {
+    if (job && this.settings.storeJobs) {
+      this.jobs.set(jobId, job);
+    }
+    return job;
+  });
+
+  if (cb) helpers.asCallback(promise, cb);
+  return promise;
 };
 
-Queue.prototype.getNextJob = function (cb) {
-  var self = this;
-  this.bclient.brpoplpush(this.toKey('waiting'), this.toKey('active'), 0, function (err, jobId) {
-    /* istanbul ignore if */
-    if (err) return cb(err);
-    return Job.fromId(self, Number(jobId), cb);
+Queue.prototype.removeJob = function (jobId, cb) {
+  const promise = this._evalScript('removeJob', 7,
+    this.toKey('succeeded'), this.toKey('failed'), this.toKey('waiting'),
+    this.toKey('active'), this.toKey('stalling'), this.toKey('jobs'),
+    this.toKey('delayed'), jobId)
+    .then(() => this);
+
+  if (cb) helpers.asCallback(promise, cb);
+  return promise;
+};
+
+Queue.prototype._waitForJob = function () {
+  const idPromise = helpers.deferred(), done = idPromise.defer();
+  this.bclient.brpoplpush(this.toKey('waiting'), this.toKey('active'), 0, done);
+
+  return idPromise.then((jobId) => {
+    // Note that the job may be null in the case that the client has removed the
+    // job before processing can take place, but after the brpoplpush has
+    // returned the job id.
+    return Job.fromId(this, jobId);
+  }, (err) => {
+    if (err.name === 'AbortError' && this.paused) {
+      return null;
+    }
+    throw err;
   });
 };
 
-Queue.prototype.runJob = function (job, cb) {
-  var self = this;
-  var psTimeout;
-  var handled = false;
+Queue.prototype._getNextJob = function () {
+  // Under normal calling conditions, commandable will not reject because we
+  // will have just checked paused in Queue#process.
+  return this._commandable(true).then(() => this._waitForJob());
+};
+
+Queue.prototype._runJob = function (job) {
+  let psTimeout = null, completed = false;
 
-  var preventStalling = function () {
-    self.client.srem(self.toKey('stalling'), job.id, function () {
-      if (!handled) {
-        psTimeout = setTimeout(preventStalling, self.settings.stallInterval / 2);
-      }
+  const preventStalling = () => {
+    psTimeout = null;
+    if (this._isClosed) return;
+    this._preventStall(job.id).then(() => {
+      if (completed || this._isClosed) return;
+      psTimeout = setTimeout(preventStalling, this.settings.stallInterval / 2);
     });
   };
   preventStalling();
 
-  var handleOutcome = function (err, data) {
-    // silently ignore any multiple calls
-    if (handled) {
-      return;
+  const handleOutcome = (err, data) => {
+    completed = true;
```

Comment:
In this case, it isn't necessary because we moved to promises. The promise cannot `resolve` and/or `reject` twice, so this function will be called at most once.